Add unit tests for StringUtils and TimeUtils

shafi-1 opened the pull request:
I've analysed your codebase and noticed that `azkaban-common/src/test/java/azkaban/utils/StringUtils` and `az-core/src/test/java/azkaban/utils/TimeUtils` are not fully tested.
I've written some tests for the methods in these classes with the help of Diffblue Cover.

Hopefully, these tests will help you detect any regressions caused by future code changes. If you would find it useful to have additional tests written for this repository, I would be more than happy to look at other classes that you consider important in a subsequent PR.

paulkennethkent commented:
@shafi-1 You need to clean up the commit history

Braavos6 commented on az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:
```diff
@@ -0,0 +1,105 @@
+package azkaban.utils;
```
Header missing

paulkennethkent replied:
Licence header

shafi-1 replied:
added license header

petersear replied:
I can't see the licence header. On a quick check, I didn't see some of the other changes either.
Is there a new commit which hasn't been pushed yet?

shafi-1 replied:
i'll check

shafi-1 replied:
should be updated now :)

Braavos6 commented on azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java:
```diff
@@ -3,78 +3,138 @@
 import org.junit.Assert;
```
Header missing

shafi-1 replied:
added license header

Braavos6 commented on az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:
```diff
@@ -0,0 +1,105 @@
+    @Test
+    public void testFormatDateTime() {
+        Assert.assertEquals("-",
+                TimeUtils.formatDateTime(-8_070_450_532_247_928_832L));
```
Add a return type which is an actual date

paulkennethkent replied:
What Eric means is that you want to pass in something that asserts a real date

shafi-1 replied:
returns an actual date now, kept the previous assertion for added coverage

Braavos6 commented on az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:
```diff
@@ -0,0 +1,105 @@
+    @Test
+    public void testFormatDateTimeZone() {
+        Assert.assertEquals("-",
+                TimeUtils.formatDateTimeZone(-8_070_450_532_247_928_832L));
```
Add a return type which is an actual date

paulkennethkent replied:
What Eric means is that you want to pass in something that asserts a real date

shafi-1 replied:
returns an actual date

Braavos6 commented on az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    @Test
+    public void testFormatDuration() {
```
branch coverage for
`    if (endTime == -1) {
      durationMS = System.currentTimeMillis() - startTime;`

is missing

shafi-1 replied:
I tried doing this, but I'd have to use mocking and it's difficult building the project when powermockito is added as it conflicts with the existing mockito dependencies

Braavos6 commented on az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    @Test
+    public void testCreatePeriodString() {
```
branch and line coverage add test case with parameter string `null` and `n`

Braavos6 commented on az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:
```diff
@@ -0,0 +1,105 @@
+        Assert.assertEquals("Months",
+                TimeUtils.parsePeriodString("12M").getPeriodType().getName());
+        Assert.assertEquals("Seconds",
+                TimeUtils.parsePeriodString("120s").getPeriodType().getName());
```
Add one test case for `IllegalArgumentException`

Braavos6 commented on az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:
```diff
@@ -0,0 +1,105 @@
+        Assert.assertEquals("Seconds",
+                TimeUtils.parsePeriodString("120s").getPeriodType().getName());
+    }
+}
```
Add blank line

petersear replied:
Still need a `carriage return` at the end of the file as mentioned by @Braavos96

petersear replied:
Still needs resolving.

shafi-1 replied:
i added a blank line at the end of the file

Braavos6 commented on az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    @Test
+    public void testParsePeriodString() throws NullPointerException {
```
Move this method one higher.. Above this `testCreatePeriodString`

shafi-1 replied:
done

Braavos6 commented on azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java:
```diff
@@ -3,78 +3,138 @@
+  }
+
+  @Test
+  public void testJoinArrayListNull() {
```
why didn't you put all the tests for `join` in one method?

shafi-1 replied:
done

Braavos6 commented on azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java:
```diff
@@ -3,78 +3,138 @@
+	list.add("!!");
+	list.add("??");
+
+	Assert.assertEquals("!!$$$??$$$", StringUtils.join(list, "$$$"));
```
you could use Arrays.asList() to do 
`	Assert.assertEquals("!!$$$??$$$",
            StringUtils.join(new ArrayList<>(Arrays.asList("!!","??")),
                    "$$$"));`

shafi-1 replied:
changed

Braavos6 commented on azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java:
```diff
@@ -3,78 +3,138 @@
+	final ArrayList<String> list = new ArrayList<>();
+	list.add("");
+
+	Assert.assertEquals("!", StringUtils.join(list, "!"));
```
same here and below

shafi-1 replied:
changed except one of them where it had to be a Collection to obtain higher coverage

Braavos6 commented on azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java:
```diff
@@ -3,78 +3,138 @@
+
+  @Test
+  public void testIsFromBrowser() {
+	Assert.assertFalse(StringUtils.isFromBrowser(null));
```
Add these for line and branch coverage
`      Assert.assertFalse(StringUtils.isFromBrowser("foo"));
      Assert.assertTrue(StringUtils.isFromBrowser("*Gecko"));
      Assert.assertTrue(StringUtils.isFromBrowser("*Chrome"));
      Assert.assertTrue(StringUtils.isFromBrowser("*Trident"));
      Assert.assertTrue(StringUtils.isFromBrowser("*AppleWebKit"));`

shafi-1 replied:
added foo but the rest oddly break the test

petersear replied:
I just tried the suggestions above and the tests passed. Might be worth another try?

shafi-1 replied:
it works, I must have missed the *

Braavos6 reviewed (changes requested):
Check the spacing for this project please

paulkennethkent commented:
I think most of what has Eric makes sense. If you don't understand certain parts then I can sit down and help.

petersear commented on az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:
```diff
@@ -0,0 +1,129 @@
+				TimeUtils.formatDuration(-4462057L, -20994L));
+		Assert.assertEquals("1290d 15h 0m",
+				TimeUtils.formatDuration(-833737520625L, -722227520546L));
+		Assert.assertEquals("7 sec",
```
The duration increases until the last assert. I think it would sense for this assert to appear above the "1m 1s" assert.
Also, the long literals here contain underscores whereas none of the others do - make these consistent.

shafi-1 replied:
in order and removed underscores

petersear commented on az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:
```diff
@@ -0,0 +1,129 @@
+	}
+
+	@Test
+	public void testFormatPeriod() {
```
The asserts in `testFormatPeriod()` should be ordered logically. I would suggest testing in the order the method under test checks the time period ie. null first then years, months, weeks, day, hours, minutes, seconds.
I think if you add a test with period of `0`, you will cover the last `return` in the method.

shafi-1 replied:
ordered, tried adding a period of `0` but broke the test, the lines in that method seem to be covered entirely anyway

petersear replied:
My mistake, that should have been "0s" which will skip all of the `if`'s to go straight to the `return`. Maybe that isn't that useful though.

petersear commented on az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:
```diff
@@ -0,0 +1,129 @@
+	}
+
+	@Test
+	public void testParsePeriodString() throws NullPointerException {
```
List the asserts in a logical order - probably the order they are checked in the method.

shafi-1 replied:
now in order

petersear commented on az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:
```diff
@@ -0,0 +1,129 @@
+	}
+
+	@Test
+	public void testParsePeriodStringException() throws IllegalArgumentException {
```
This test should follow testParsePeriodString() rather than testCreatePeriodString()

shafi-1 replied:
swapped

petersear commented on az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:
```diff
@@ -0,0 +1,129 @@
+	}
+
+	@Test
+	public void testCreatePeriodString() {
```
Order the asserts logically.

shafi-1 replied:
ordered

petersear commented on azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java:
```diff
@@ -1,80 +1,138 @@
 public class StringUtilsTest {
 
   private static final String chromeOnMac =
-      "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/44.0.2403.155 Safari/537.36";
```
There seem to be a lot of whitespace changes here (spaces have become tabs?)

shafi-1 replied:
I think I accidentally changed the spaces to tabs, switched back to spaces

petersear commented on azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java:
```diff
@@ -1,80 +1,138 @@
+
+  @Test
+  public void testShellQuote() {
+	Assert.assertEquals("!\\!!", StringUtils.shellQuote("!", '!'));
```
You could use more meaningful inputs eg. a readable string.

shafi-1 replied:
tried to add more meaningful inputs

petersear commented on azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java:
```diff
@@ -1,80 +1,138 @@
+			new ArrayList<>(Arrays.asList("")), "!"));
+	Assert.assertEquals("", StringUtils.join(
+			new ArrayList<>(), null));
+	Assert.assertEquals("!!$$$??$$$", StringUtils.join(
```
Could use more meaningful inputs (optional)

shafi-1 replied:
same here

petersear commented on build.gradle:
```diff
@@ -125,6 +125,7 @@ ext.deps = [
         jacksonDatabind     : 'com.fasterxml.jackson.core:jackson-databind:2.7.4',
         jasypt              : 'org.jasypt:jasypt:1.9.2',
         jacksonCore         : 'com.fasterxml.jackson.core:jackson-core:2.9.2'
+
```
Should this change be here?
If you are not modifying any lines in the file, I wouldn't expect to see a blank line added.

shafi-1 replied:
removed the blank line

petersear reviewed (changes requested):
It looks like tabs are not used in these test classes. It would be worth checking that all indents consist of the same number of spaces as the existing files rather than tabs.

petersear commented on azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java:
```diff
@@ -77,4 +97,47 @@ public void containsBrowserName() {
+  public void testIsFromBrowser() {
+	Assert.assertFalse(StringUtils.isFromBrowser(null));
+	Assert.assertFalse(StringUtils.isFromBrowser("foo"));
+	Assert.assertTrue(StringUtils.isFromBrowser("*Gecko"));
```
Insert a blank line between the different assert types (`False` and `True`)

shafi-1 replied:
inserted

petersear commented on az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:
```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2017 LinkedIn Corp.
```
They might want this updated to `2019`. On the other hand, I had feedback not to change the year so maybe just be aware.

shafi-1 replied:
noted

petersear reviewed (changes requested):
There are still some tabs in both files although it's looking better :-)
The style seems to be 2-space indents in the existing (and other) test class so worth maybe changing your IDE settings accordingly and matching the project.

petersear commented on az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:
```diff
@@ -0,0 +1,131 @@
+	}
+
+	@Test
+	public void testParsePeriodStringThrowsException() {
```
There is another test for an exception thrown by parsePeriodString() below testParsePeriodString(), which is the right location for it, so this one could be removed I think?

shafi-1 replied:
removed the extra test

petersear commented on az-core/src/test/java/azkaban/utils/TimeUtilsTest.java:
```diff
@@ -0,0 +1,131 @@
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+public class TimeUtilsTest {
```
This whole (new) test class is indented using tabs but should be indented using 2 spaces to match other existing test files.

shafi-1 replied:
managed to change

petersear commented on azkaban-common/src/test/java/azkaban/utils/StringUtilsTest.java:
```diff
@@ -1,8 +1,28 @@
+import java.util.Arrays;
+import java.util.Collection;
+
 public class StringUtilsTest {
```
The existing tests in this class use 2 spaces for indents (and 4 for a continuation line)
The new tests use an 8-space tab which is equivalent to indenting by 6 spaces relative to the previous line.
This should be changed so that the style of the added tests matches that of the existing tests.

shafi-1 replied:
changed to match the style

shafi-1 committed:
Add unit tests for StringUtils and TimeUtils

Fully-qualified classnames:

azkaban-common/src/test/java/azkaban/utils/StringUtils
az-core/src/test/java/azkaban/utils/TimeUtils

These tests were written using Diffblue Cover.
